workflow/controller/operator.go

```diff
@@ -2175,54 +2122,70 @@ func (woc *wfOperationCtx) markNodeWaitingForLock(nodeName string, lockName stri
 	if woc.execWf.Spec.Parallelism != nil && woc.activePods >= *woc.execWf.Spec.Parallelism {
 		woc.log.Infof("workflow active pod spec parallelism reached %d/%d", woc.activePods, *woc.execWf.Spec.Parallelism)
-		return ErrParallelismReached
+		return node, ErrParallelismReached
 	}
 	// TODO: repeated calls to countActivePods is not optimal
 	switch tmpl.GetType() {
 	case wfv1.TemplateTypeDAG, wfv1.TemplateTypeSteps:
-		// if we are about to execute a DAG/Steps template, make sure we havent already reached our limit
-		if tmpl.Parallelism != nil && node != nil {
-			templateActivePods := woc.countActivePods(node.ID)
-			if templateActivePods >= *tmpl.Parallelism {
+		if node != nil {
+			// if we are about to execute a DAG/Steps template, make sure we havent already reached our limit
+			counts := woc.countNodes(getActivePodsCounter(node.ID), getFailedOrErroredChildrenCounter(node.ID))
+			templateActivePods := int64(counts.getCountType(counterTypeActivePods))
+			templateFailedOrErroredChildren := int64(counts.getCountType(counterTypeFailedOrErroredChildren))
+
+			if tmpl.FailFast != nil && *tmpl.FailFast && templateFailedOrErroredChildren > 0 {
+				woc.markNodePhase(node.Name, wfv1.NodeFailed, "template has failed or errored children and failFast enabled")
+				return node, ErrParallelismReached
+			}
+
+			if tmpl.Parallelism != nil && templateActivePods >= *tmpl.Parallelism {
 				woc.log.Infof("template (node %s) active pod parallelism reached %d/%d", node.ID, templateActivePods, *tmpl.Parallelism)
-				return ErrParallelismReached
+				return node, ErrParallelismReached
 			}
 		}
+
 		fallthrough
 	default:
-		// if we are about to execute a pod, make our parent hasn't reached it's limit
+		// if we are about to execute a pod, make sure our parent hasn't reached it's limit
 		if boundaryID != "" && (node == nil || (node.Phase != wfv1.NodePending && node.Phase != wfv1.NodeRunning)) {
 			boundaryNode, ok := woc.wf.Status.Nodes[boundaryID]
 			if !ok {
-				return errors.InternalError("boundaryNode not found")
+				return node, errors.InternalError("boundaryNode not found")
 			}
 			tmplCtx, err := woc.createTemplateContext(boundaryNode.GetTemplateScope())
 			if err != nil {
-				return err
+				return node, err
 			}
 			_, boundaryTemplate, templateStored, err := tmplCtx.ResolveTemplate(&boundaryNode)
 			if err != nil {
-				return err
+				return node, err
 			}
 			// A new template was stored during resolution, persist it
 			if templateStored {
 				woc.updated = true
 			}
 
 			if boundaryTemplate != nil && boundaryTemplate.Parallelism != nil {
-				activeSiblings := woc.countActiveChildren(boundaryID)
+				counts := woc.countNodes(getActiveChildrenCounter(boundaryID), getFailedOrErroredChildrenCounter(boundaryID))
+				activeSiblings := int64(counts.getCountType(counterTypeActiveChildren))
+				templateFailedOrErroredChildren := int64(counts.getCountType(counterTypeFailedOrErroredChildren))
```

Comment:
“FailedOrErrored” is wordy, how about “unsuccessfully”?

Reply:
These could be expensive (O(N x 3) ?) operations. But only needed when you have FailFast? Maybe guard them this that flag.